pulsar-common/src/main/java/org/apache/pulsar/common/util/RateLimiter.java

```diff
@@ -257,14 +267,14 @@ public synchronized TimeUnit getRateTimeUnit() {
     }
 
     synchronized void renew() {
-        acquiredPermits = isDispatchRateLimiter ? Math.max(0, acquiredPermits - permits) : 0;
+        acquiredPermits = isDispatchOrPrecisePublishRateLimiter ? Math.max(0, acquiredPermits - permits) : 0;
         if (permitUpdater != null) {
             long newPermitRate = permitUpdater.get();
             if (newPermitRate > 0) {
                 setRate(newPermitRate);
             }
         }
-        if (rateLimitFunction != null) {
+        if (rateLimitFunction != null && this.getAvailablePermits() > 0) {
```

Comment:
This seems to make sense. Please add some comments here explaining the logic.

Reply:
I'd assume that the condition would need to handle the case where `isDispatchOrPrecisePublishRateLimiter` is `false`. Did you think about that?

Reply:
Yea Ill add a comment.

And actually i didn't give it much thought but when I'm thinking about it the rateLimitFunction is a callback that lets the outer scope access to the renew function, I don't think that we should use this property without knowing exactly what expected to be happening.

I believe that checking whether there are available permits is the reasonable condition here because we would want to let the outer scope a way to run something when there are any available permits, and it doesn't really depend on the class property.

If it sets to false we can assume the user of the rateLimiter wants the state to be reset every time window otherwise he probably want back-pressuring something.

Reply:
Done